qiita_db/metadata_template/prep_template.py

```diff
@@ -713,3 +713,192 @@ def modification_timestamp(self):
     @staticmethod
     def max_samples():
         return qdb.util.max_preparation_samples()
+
+    def add_default_workflow(self, user):
+        """The modification timestamp of the prep information
+
+
+        Parameters
+        ----------
+        user : The user that requested to add the default workflows
+
+        Returns
+        -------
+        ProcessingWorkflow
+            The workflow created
+
+        Raises
+        ------
+        ValueError
+            If this preparation doesn't have valid workflows
+            If this preparation has been fully processed
+        """
+        # helper functions to avoid duplication of code
+
+        def _get_node_info(node):
+            # retrieves the merging scheme of a node
+            parent = list(wk.graph.predecessors(node))
+            if parent:
+                parent = parent.pop()
+                pdp = parent.default_parameter
+                pcmd = pdp.command
+                pparams = pdp.values
+            else:
+                pcmd = None
+                pparams = {}
+
+            dp = node.default_parameter
+            cparams = dp.values
+            ccmd = dp.command
+
+            parent_cmd_name = None
+            parent_merging_scheme = None
+            if pcmd is not None:
+                parent_cmd_name = pcmd.name
+                parent_merging_scheme = pcmd.merging_scheme
+
+            return qdb.util.human_merging_scheme(
+                ccmd.name, ccmd.merging_scheme, parent_cmd_name,
+                parent_merging_scheme, cparams, [], pparams)
+
+        def _get_predecessors(node):
+            # recursive method to get predecessors of a given node
+            for pnode in wk.graph.predecessors(node):
+                pred = _get_predecessors(pnode)
+                cxns = {x[0]: x[2]
+                        for x in wk.graph.get_edge_data(
+                            pnode, node)['connections'].connections}
+                data = [pnode, node, cxns]
+                if pred is None:
+                    pred = [data]
+                else:
+                    pred.append(data)
+                return pred
```

Comment:
I *think* that having the `return` in scope of the `for` would imply a node can only have a single predecessor. Is that correct for both interpretation and intent?

Reply:
Yes!! Note that this is a recursive function (line 767 calls itself) as we need the "grandparent" information to create the "parent" information.

Reply:
Ya, but just so I understand, is it correct that a node only ever has a single parent?

Reply:
Thank you for double checking this; better to be safe. 

Yes, I spent sometime looking at this and this is the expected behavior; basically, by being recursive it will go down to the origin of the workflow (always a single artifact) and return (`if pred is None:`) the single only parent, then it will start coming out of the recursion and appending in order the children jobs (`else:`). In case it helps, we tested this in qiita-rc and everything seem fine.